lib/http2/stream.c

```diff
@@ -352,8 +353,12 @@ void finalostream_send(h2o_ostream_t *self, h2o_req_t *req, h2o_sendvec_t *bufs,
             return;
         }
         h2o_probe_log_response(&stream->req, stream->stream_id);
-        if (send_headers(conn, stream) != 0)
+        if (send_headers(conn, stream, is_end_stream) != 0)
             return;
+        if (is_end_stream) {
+            h2o_http2_stream_set_state(conn, stream, H2O_HTTP2_STREAM_STATE_END_STREAM);
+            break;
```

Comment:
By breaking here, we will be calling `h2o_http2_conn_register_for_proceed_callback`. Does not that lead to us sending a zero-sized DATA frame with the END_STREAM flag?

I know that the test does not detect such a behavior but wonder if that is just due to how nghttp works (i.e., close the connection before processing any extraneous frames that might follow).

If that is the case, maybe doing something like `send_refused_stream` called right above might make more sense.

Reply:
Thanks for catching that, i've updated the test to catch the case and fixed the code in 5507e3d